lib/realtime/vehicle.ex

```diff
@@ -64,18 +76,103 @@ defmodule Realtime.Vehicle do
   """
   @spec decode(term()) :: t()
   def decode(%{} = json) do
+    stop_times_on_trip_fn =
+      Application.get_env(:realtime, :stop_times_on_trip_fn, &Gtfs.stop_times_on_trip/1)
+
+    id = json["id"]
+    label = json["vehicle"]["vehicle"]["label"]
+    timestamp = json["vehicle"]["timestamp"]
+    direction_id = json["vehicle"]["trip"]["direction_id"]
+    route_id = json["vehicle"]["trip"]["route_id"]
+    trip_id = json["vehicle"]["trip"]["trip_id"]
+    current_stop_status = decode_current_status(json["vehicle"]["current_status"])
+    stop_id = json["vehicle"]["stop_id"]
+
+    stop_times_on_trip =
+      try do
+        stop_times_on_trip_fn.(trip_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    timepoint_status =
+      with {percent_of_the_way_to_next_timepoint, next_timepoint_stop_time} <-
+             percent_of_the_way_to_next_timepoint(stop_times_on_trip, stop_id) do
+        %{
+          status:
+            current_timepoint_status(current_stop_status, next_timepoint_stop_time, stop_id),
+          timepoint_id: next_timepoint_stop_time && next_timepoint_stop_time.timepoint_id,
+          percent_of_the_way_to_timepoint: percent_of_the_way_to_next_timepoint
+        }
+      else
+        nil ->
+          nil
+      end
+
     %__MODULE__{
-      id: json["id"],
-      label: json["vehicle"]["vehicle"]["label"],
-      timestamp: json["vehicle"]["timestamp"],
-      direction_id: json["vehicle"]["trip"]["direction_id"],
-      route_id: json["vehicle"]["trip"]["route_id"],
-      trip_id: json["vehicle"]["trip"]["trip_id"],
-      current_status: decode_current_status(json["vehicle"]["current_status"]),
-      stop_id: json["vehicle"]["stop_id"]
+      id: id,
+      label: label,
+      timestamp: timestamp,
+      direction_id: direction_id,
+      route_id: route_id,
+      trip_id: trip_id,
+      stop_status: %{
+        status: current_stop_status,
+        stop_id: stop_id
+      },
+      timepoint_status: timepoint_status
     }
   end
 
+  @spec percent_of_the_way_to_next_timepoint([StopTime.t()], Stop.id()) ::
+          {non_neg_integer(), StopTime.t() | nil} | nil
+  def percent_of_the_way_to_next_timepoint([], _stop_id), do: {0, nil}
+
+  def percent_of_the_way_to_next_timepoint(stop_times, stop_id) do
+    {past_stop_times, future_stop_times} = Enum.split_while(stop_times, &(&1.stop_id != stop_id))
+
+    next_timepoint_stop_time = Enum.find(future_stop_times, &is_a_timepoint?(&1))
```

Comment:
I thought we'd discussed cases where a route didn't have timepoints (at least in sections?), and so we were treating the final stop as the next timepoint. Am I confusing things?

Reply:
Oh yeah, I had forgotten about that case.

My current opinion is to lean slightly toward making the whole thing `nil` in that case, because without a timepoint id, it'd be difficult to do anything useful with the percentage on the frontend. But I'd be fine either way.

Reply:
I think it's the opposite, I think this lets you do something with it on the frontend. With this info you can plot progress towards the end of the line. You couldn't do that if the whole struct was nil.